### ISSUE TITLE
[Questions] Coil re-entrant calls and cache without OkHTTP

### ISSUE THREAD

**Issue by Tolriq** [coil-kt/coil#305]
Sorry for the question title but can't find a better one.

I plan to test this library to replace Glide, but I have a relatively quite complex needs that had problems with Glide.
The app require an infinite cache of high quality large images that are used later at different sizes, those image can be loaded as the user browse or he can force some preloading.

The first question is related to what was an issue with Glide: Does Coil support re-entrant calls inside a custom fetcher? (The app calls load(data) the mapper calls the custom fetcher that will in turn use Coil to load from the network as a suspend to have the image at another size saved to it's private cache.

The second question is about final images caching, reading some issues it seems Coil does not offer disk cache for transformed result images as only support untouched source image caching via OkHTTP cache. This means that since I can have source images quite large but then loaded at really smaller sizes in some recyclerview for example the read times would be way too high. Have this changed and not documented yet?

**Comment by colinrtwhite** [comment]
Hmm I'm not sure I completely understand the first question. It's not possible to have a custom fetcher call Coil's `HttpFetcher`, as only one fetcher is called per request. Sounds like you might want a `MeasuredMapper` to await the current size of the target? You can then use the size to decide what URL to get from your server.

Yep, Coil does not cache transformed images on disk - only in memory. That said, loading large images at small sizes should still be quite quick. Coil uses a number of optimizations (the same optimizations Glide uses) to decode large images into memory.

**Comment by Tolriq** [comment]
My application can connect to many different media server that will serve many different kind of images with resolution up to 4K and no way to have them resized at server side for some of them.

Those images are then displayed at many different size depending on screen and user configuration and the application must work offline first with complete image cache. (Like small image in a grid then full screen image)

So the way it's handled is that the application makes normal request with a custom data object, the fetcher returns either it's local cache or download the image, resize it at the maximum size it can be displayed store it in cache and return that image.

The goal would be to always return from cache but being able to use Coil to make the network request + resize and not do it manually like I do with Glide. If it's not possible then it's still a lot of work done manually :(

For the loading from disk of large images vs small ones + do the resize even with glide I can assure you than the difference is large enough for low end to medium end devices to completely worth it during fast scrolling.

Under some configuration like a tablet and user selecting small grid, there can be 75 to 100 or even more images displayed per screen in a long recyclerview. Even with bitmap pools, insample and all optimisations, loading direct 150px images at correct ratio, vs loading 1080px images that require resize impact is insane difference.

**Comment by colinrtwhite** [comment]
Hmm if I'm understanding correctly it sounds like you might have to write something custom for this. Coil is backed by OkHttp's disk cache and using different URLs will be saved as different entries in the cache. I'd check out the `MeasuredMapper` and `Fetcher` classes and see if that fits your use case. At the least, the code you've written for Glide is likely to be translatable to Coil.

**Comment by Tolriq** [comment]
Yes I need to do manual things :) But I'd like to still embrace the library, because if I can't reuse coil suspend get() function (so "waiting" call) inside the custom fetcher and can't cache the final transformation this means that I need to write an http fetcher + resizer + cache handling + transformed cache handling ;)
This is about 80% of an image library job :( So in the end I could also write the rest.

About code reuse, not really, many things are really specific to Glide + reuse some of their code like the rewinding buffered reading. (And I've contributed a few things to Glide to be able to do what I needed). So in the end even if I'd like to drop Glide for a full coroutine way I'm not sure I have enough time to rebuild everything and try to get some PR merged in Coil :(

I've seen some of your messages about starting to profile Coil, you may really want to profile the benefits of transformed image cache that may make you reconsider caching inside Coil (At least as an optional component)

**Comment by colinrtwhite** [comment]
I wrote up some thoughts [here](https://github.com/coil-kt/coil/issues/113#issuecomment-533842984) on why I don't think caching images post-transformation is a good fit for Coil.

**Comment by Tolriq** [comment]
Yes I've read that and that's why I asked here if things changed since then :'(

Many apps does not only connect to controlled servers, my app connecting to many different media center that won't allow transformation + non proper header is an example.
If the server serves 4K PNG images and we only rely on OkHTTP cache this means that the 250Mb cache will rotate fast when the user have large collection of media. So there will be tons of downloads of very large images + when loading from disk to display a 150px* 150px image we'll need to load that image with inSample that won't match precisely + do the resize this is slow and consume lot's of memory. (And I don't know how you handle the bitmap pool but if all images displayed are at fixed size it's pretty efficient but reusing bitmaps from inSampleSize is way more harder as the source image size will often be different and the resulting size is always multiple of the original one, leading to more miss, more eviction and more allocations)

Anyway I completely understand why you may not want to address all the cases and make this a specialized optimized library relying on external mechanisms. But this means that I'm stuck with Glide for the moment :'(

**Comment by colinrtwhite** [comment]
> And I don't know how you handle the bitmap pool but if all images displayed are at fixed size it's pretty efficient but reusing bitmaps from inSampleSize is way more harder as the source image size will often be different and the resulting size is always multiple of the original one, leading to more miss, more eviction and more allocations

Coil will reuse bitmaps for the image's size in memory (not the source dimensions) which should be efficient in a list of small images.

Thinking about this more, you might be able to achieve this functionality by using an [`OkHttp Interceptor`](https://square.github.io/okhttp/interceptors/) to rewrite the network request's response body. This would allow you to resize the image in memory before it is saved in the disk cache.

Going to close this out since I think the initial question has been answered.

**Comment by Tolriq** [comment]
@colinrtwhite now that there's the listener API I made a quick test to measure things for my large image in cache size requirement.

Using the sample with full images more or less represent my need (Those numbers can be reduced by using an intermediary cache where I would resize the image to lower size, but those size would still be high to allow full quality full screen images without loss) and the result are corresponding to my expectations (on a Pixel 3a so fart from low end device)

```
2020-03-19 15:17:15.947 11349-11489/coil.sample E/Coil: Fetch duration 105 for https://images.unsplash.com/photo-1550900167-d1cc07eec3d7?ixlib=rb-1.2.1&q=85&fm=jpg&crop=entropy&cs=srgb&ixid=eyJhcHBfaWQiOjU4MjM5fQ
2020-03-19 15:17:17.021 11349-11489/coil.sample E/Coil: Decode duration 1180 for https://images.unsplash.com/photo-1550900167-d1cc07eec3d7?ixlib=rb-1.2.1&q=85&fm=jpg&crop=entropy&cs=srgb&ixid=eyJhcHBfaWQiOjU4MjM5fQ
2020-03-19 15:17:17.021 11349-11489/coil.sample E/Coil: Transform duration 0 for https://images.unsplash.com/photo-1550900167-d1cc07eec3d7?ixlib=rb-1.2.1&q=85&fm=jpg&crop=entropy&cs=srgb&ixid=eyJhcHBfaWQiOjU4MjM5fQ
2020-03-19 15:17:17.022 11349-11349/coil.sample I/RealImageLoader: 💾 Successful (DISK) - https://images.unsplash.com/photo-1550900167-d1cc07eec3d7?ixlib=rb-1.2.1&q=85&fm=jpg&crop=entropy&cs=srgb&ixid=eyJhcHBfaWQiOjU4MjM5fQ
2020-03-19 15:17:17.133 11349-11349/coil.sample E/Coil: Load duration 1292 for https://images.unsplash.com/photo-1550900167-d1cc07eec3d7?ixlib=rb-1.2.1&q=85&fm=jpg&crop=entropy&cs=srgb&ixid=eyJhcHBfaWQiOjU4MjM5fQ

```

The decode duration is wayyy too long on those kind of large image and the reason why a processed image cache like on Glide is important for quite a few needs :(
And this is without transformation cost that could be added.

I'll stop bothering you but hope your benchmarks will measure this for the comparisons and that maybe that opens an investigation on an optional API to allow that. (Request have keys and alias keys, transformations have keys too that yes would require update when implementation change but people are used to that, even in your memory cache if the consumer reuse the same transformation key for different things there's issues).
In terms of API could an optional resultCacheFetcher with a get an a put without a default one provided by coil :) But entry points for others. Quickly reading code it's just 2 calls inside execute of RealImageLoader but as everything is internal / private it's currently not possible to easily add this.

**Comment by colinrtwhite** [comment]
@Tolriq Thanks for the performance numbers. As I mentioned I don't think it's straightforward to add an API to Coil for this. Coil relies on OkHttp's cache, which doesn't allow us to make transformations to the image between network -> disk cache.

The best way to accomplish transforming the image before it's cached is to add an `Interceptor` to your `OkHttpClient` instance and transform the ResponseBody as it's being downloaded. Please test it out and let me know how it goes.

**Comment by Tolriq** [comment]
@colinrtwhite I can handle the reducing of the images with a fetcher it's not the issue here.

The issue is about a cache of the transformed image like you have a memory cache, even if I reduce the 4k images to 2k images for high res tablet this is still 2k images to load from disk cache then resize then apply transformations then display in lists.
When using the sample with full images memory cache keeps nearly nothing, so there's disk load / resize a lot with 1 second delay to display images.
If there was another layer of cache of the transformed image (so already resized + associated transformation) the performance would be a million times better.

About the API it's basically the same interface as the Memory cache, you can leave the implementation outside of Coil for other users to implement the way they want.

After the loadFrom memory cache you call that interface to load from usercache if not in memory and provider present with the exact same tests and everything after.

Then after the memory cache update you call that interface to save to usercache if a provider is present.

TL;DR an interface with a put a get, 2 calls in the RealImageLoader and a new field to configure the cache provider. This is really simple API (IMO) and allows consumer to provider more things that currently as there's no way to add this features due to current internal / private things.
